File: packages/javascript/src/retained-rectangle-cuts.js
const MAX_LEAF_ID = 9007199254740990;
const EXHAUSTED_NEXT_ID = 9007199254740991;

/** Validation/access error with a stable catalog code. */
export class RectangleCutError extends Error {
  constructor(code) {
    super(code);
    this.name = "RectangleCutError";
    this.code = code;
  }
}

function invalidInput() { throw new RectangleCutError("INVALID_INPUT"); }
function unknownId() { throw new RectangleCutError("UNKNOWN_ID"); }
function invalidCut() { throw new RectangleCutError("INVALID_CUT"); }
function limitExceeded() { throw new RectangleCutError("LIMIT_EXCEEDED"); }

function finite(value) { return typeof value === "number" && Number.isFinite(value); }
function zero(value) { return value === 0 ? 0 : value; }

function coordinate(value) {
  if (typeof value !== "number") invalidInput();
  if (!finite(value)) invalidInput();
  return zero(value);
}

function checkedId(id) {
  if (typeof id !== "number" || !Number.isInteger(id) || id < 0 || id > MAX_LEAF_ID) {
    throw new RectangleCutError("INVALID_ID");
  }
  return id;
}

function leafValue(id, left, top, right, bottom) {
  return Object.freeze({ id, bounds: Object.freeze([left, top, right, bottom]) });
}

/**
 * Caller-directed retained rectangular regions with stable never-reused identities.
 * Implements layout.retained-rectangle-cuts-2d 0.1.0.
 *
 * Independently specified from survey/out/2019/generativos/griton/notes.md. The
 * motivating work supplies retained unequal regions; its source-specific selection,
 * ratios, omission, drawing, and random behavior are outside this operation.
 */
export class RetainedRectangleCuts2D {
  #leaves = new Map();
  #nextId = 1;

  constructor(left, top, right, bottom) {
    this.#leaves.set(0, leafValue(0, left, top, right, bottom));
    Object.seal(this);
  }

  /** Replaces a live leaf with low-coordinate then high-coordinate children; returns their IDs. */
  cut(id, axis, coord) {
    checkedId(id);
    const parent = this.#leaves.get(id);
    if (parent === undefined) unknownId();
    if (axis !== "X" && axis !== "Y") invalidInput();
    if (typeof coord !== "number" || !finite(coord)) invalidInput();
    const c = zero(coord);
    const cutX = axis === "X";
    const [pl, pt, pr, pb] = parent.bounds;
    const low = cutX ? pl : pt;
    const high = cutX ? pr : pb;
    if (!(c > low && c < high)) invalidCut();
    if (this.#leaves.size >= 2147483647 || this.#nextId > EXHAUSTED_NEXT_ID - 2) limitExceeded();

    const lowId = this.#nextId, highId = this.#nextId + 1;
    const lowLeaf = cutX ? leafValue(lowId, pl, pt, c, pb) : leafValue(lowId, pl, pt, pr, c);
    const highLeaf = cutX ? leafValue(highId, c, pt, pr, pb) : leafValue(highId, pl, c, pr, pb);
    this.#leaves.delete(id);
    this.#leaves.set(lowId, lowLeaf);
    this.#leaves.set(highId, highLeaf);
    this.#nextId += 2;
    return [lowId, highId];
  }

  /** Removes a live leaf without replacing it or changing later identity allocation. */
  remove(id) {
    checkedId(id);
    if (!this.#leaves.delete(id)) unknownId();
  }

  /** Returns the immutable value for a live leaf. */
  leaf(id) {
    checkedId(id);
    const leaf = this.#leaves.get(id);
    if (leaf === undefined) unknownId();
    return leaf;
  }

  /** Returns a detached array in current live-leaf order. */
  leaves() { return Array.from(this.#leaves.values()); }

  /** Returns the current live-leaf count. */
  get size() { return this.#leaves.size; }

  /**
   * Materializes a fully detached mutable {nextId,leaves:[{id,bounds}]} snapshot.
   * This snapshot is descriptive only and is not a restoration or replay input.
   */
  toValues() {
    const leaves = [];
    for (const leaf of this.#leaves.values()) leaves.push({ id: leaf.id, bounds: leaf.bounds.slice() });
    return { nextId: this.#nextId, leaves };
  }
}

/** Creates a root from exactly {bounds:[left,top,right,bottom]}. */
export function retainedRectangleCuts2D(config) {
  if (config === null || typeof config !== "object" || Array.isArray(config)) invalidInput();
  const prototype = Object.getPrototypeOf(config);
  if (prototype !== Object.prototype && prototype !== null) invalidInput();
  if (Reflect.ownKeys(config).length !== 1 || !Object.prototype.hasOwnProperty.call(config, "bounds")) invalidInput();
  const rawBounds = config.bounds;
  if (!Array.isArray(rawBounds) || Object.getPrototypeOf(rawBounds) !== Array.prototype || rawBounds.length !== 4) {
    invalidInput();
  }
  const left = coordinate(rawBounds[0]);
  const top = coordinate(rawBounds[1]);
  const right = coordinate(rawBounds[2]);
  const bottom = coordinate(rawBounds[3]);
  return createValidated(left, top, right, bottom);
}

function createValidated(left, top, right, bottom) {
  if (!(left < right) || !(top < bottom)) invalidInput();
  const width = right - left, height = bottom - top;
  if (!finite(width) || !finite(height)) invalidInput();
  return new RetainedRectangleCuts2D(left, top, right, bottom);
}

File: packages/python/procedurals/retained_rectangle_cuts.py
"""Caller-directed retained rectangular regions with stable never-reused identities.

Implements layout.retained-rectangle-cuts-2d 0.1.0. Independently specified from
survey/out/2019/generativos/griton/notes.md. The motivating work supplies retained
unequal regions; its source-specific selection, ratios, omission, drawing, and random
behavior are outside this operation.
"""
import math
from types import MappingProxyType

__all__ = ["RectangleCutError", "RetainedRectangleCuts2D", "retained_rectangle_cuts_2d"]

_MAX_LEAF_ID = 9007199254740990
_EXHAUSTED_NEXT_ID = 9007199254740991


class RectangleCutError(ValueError):
    """Stable catalog code for retained-rectangle-cuts-2d."""
    def __init__(self, code):
        super().__init__(code)
        self.code = code


def _invalid_input():
    raise RectangleCutError("INVALID_INPUT")


def _unknown_id():
    raise RectangleCutError("UNKNOWN_ID")


def _invalid_cut():
    raise RectangleCutError("INVALID_CUT")


def _limit_exceeded():
    raise RectangleCutError("LIMIT_EXCEEDED")


def _zero(value):
    return 0.0 if value == 0.0 else value


def _coordinate(value):
    if type(value) not in (int, float):
        _invalid_input()
    value = float(value)
    if not math.isfinite(value):
        _invalid_input()
    return _zero(value)


def _checked_id(identity):
    if type(identity) is not int or isinstance(identity, bool) or not 0 <= identity <= _MAX_LEAF_ID:
        raise RectangleCutError("INVALID_ID")
    return identity


def _leaf_value(identity, left, top, right, bottom):
    return MappingProxyType({"id": identity, "bounds": (left, top, right, bottom)})


class RetainedRectangleCuts2D:
    """Mutable retained rectangle-cut model. Not thread-safe; single-writer only."""
    __slots__ = ("_leaves", "_next_id")

    def __init__(self, left, top, right, bottom):
        self._leaves = {0: _leaf_value(0, left, top, right, bottom)}
        self._next_id = 1

    def cut(self, identity, axis, coord):
        """Replace a live leaf with low-coordinate then high-coordinate children."""
        _checked_id(identity)
        parent = self._leaves.get(identity)
        if parent is None:
            _unknown_id()
        if axis not in ("X", "Y"):
            _invalid_input()
        if type(coord) not in (int, float):
            _invalid_input()
        coord = float(coord)
        if not math.isfinite(coord):
            _invalid_input()
        coord = _zero(coord)
        cut_x = axis == "X"
        pl, pt, pr, pb = parent["bounds"]
        low = pl if cut_x else pt
        high = pr if cut_x else pb
        if not (coord > low and coord < high):
            _invalid_cut()
        if len(self._leaves) >= 2147483647 or self._next_id > _EXHAUSTED_NEXT_ID - 2:
            _limit_exceeded()

        low_id, high_id = self._next_id, self._next_id + 1
        if cut_x:
            low_leaf = _leaf_value(low_id, pl, pt, coord, pb)
            high_leaf = _leaf_value(high_id, coord, pt, pr, pb)
        else:
            low_leaf = _leaf_value(low_id, pl, pt, pr, coord)
            high_leaf = _leaf_value(high_id, pl, coord, pr, pb)
        del self._leaves[identity]
        self._leaves[low_id] = low_leaf
        self._leaves[high_id] = high_leaf
        self._next_id += 2
        return [low_id, high_id]

    def remove(self, identity):
        """Remove a live leaf without replacing it or changing later identity allocation."""
        _checked_id(identity)
        if self._leaves.pop(identity, None) is None:
            _unknown_id()

    def leaf(self, identity):
        """Return the immutable value for a live leaf."""
        _checked_id(identity)
        leaf = self._leaves.get(identity)
        if leaf is None:
            _unknown_id()
        return leaf

    def leaves(self):
        """Return a detached list in current live-leaf order."""
        return list(self._leaves.values())

    @property
    def size(self):
        return len(self._leaves)

    def to_values(self):
        """Materialize a fully detached mutable {nextId,leaves:[{id,bounds}]} snapshot."""
        leaves = [{"id": leaf["id"], "bounds": list(leaf["bounds"])} for leaf in self._leaves.values()]
        return {"nextId": self._next_id, "leaves": leaves}


def _create_validated(left, top, right, bottom):
    if not (left < right) or not (top < bottom):
        _invalid_input()
    width, height = right - left, bottom - top
    if not math.isfinite(width) or not math.isfinite(height):
        _invalid_input()
    return RetainedRectangleCuts2D(left, top, right, bottom)


def retained_rectangle_cuts_2d(config):
    """Create a root from exactly {"bounds": [left, top, right, bottom]}."""
    if type(config) is not dict or len(config) != 1 or "bounds" not in config:
        _invalid_input()
    raw_bounds = config["bounds"]
    if type(raw_bounds) is not list or len(raw_bounds) != 4:
        _invalid_input()
    left = _coordinate(raw_bounds[0])
    top = _coordinate(raw_bounds[1])
    right = _coordinate(raw_bounds[2])
    bottom = _coordinate(raw_bounds[3])
    return _create_validated(left, top, right, bottom)
